### DISCUSSION TITLE
Störung und Brennerstörung jemals ungleich 0?

### DISCUSSION THREAD

**Discussion by empty88** [empty88/OptoLink#3] [General]
@gitzila
> Aber sag mal: Hast Du bei den Adressen für Störung oder Brennerstörung jemals andere Werte als 0 bekommen? Ich bekomme aus
der Schnittstelle immer nur "0". Habe das nun so gelöst, dass ich mir aus dem Fehlerprotokoll den letzten Eintrag merke (un im Flash sichere) und dann alle 30 Sekunden mit der aktuellen Liste aus dem Vito abgleiche. Gibt es dort einen neueren Eintrag, dann setze ich mir ein eigenes Fehlerflag und schicke mir per Telegram Bot eine Nachricht auf's Handy. Funktioniert super.
> 
> Und eine andere Sache noch: Taktet Deine Heizung auch so furchtbar oft? Wenn es nicht gerade -10 Grad hat, dann springt die Heizung alle 2 Minuten für 20 Sekunden an und schaltet dann ab, weil der Kessel zu warm wird. Ich habe das so gelöst und überhaupt erst deswegen das Gebastel mit dem ESP und dem Optolink angefangen, dass ich alle 2 Sekunden den Brennerstatus abfrage, bei Start dann die Raum-Soll RT auf 37 Grad hochsetze und nach 1 Minute wieder auf Normalwert zurück. In der einen Minute kommt dann genug kühles Wasser aus dem Haus nach und der Brenner läuft dann schön 10-20 Minuten durch. Grob gepeilt schont das nicht nur die ganze Anlage (statt 30 Starts in der Stunde nun 25 am Tag) sondern spart auch Gas. Bei den eisigen -15 Grad neulich waren wir hier bei unter 7 m³ am Tag.
> 
> Falls Du ein ähnliches Problem mit Deiner Heizung hast, können wir uns gern austauschen.
> 
> LG
> Frank

**Comment by empty88** [comment]
@gitzila
Hi Frank,
bei mir sieht die Taktung gerade so aus, also definitiv nicht so oft:
![image](https://user-images.githubusercontent.com/29315283/210087920-b72c06de-717f-4ff7-9ebc-f12dc63be25b.png)

Eine 1 bei Störung hatte ich noch nicht, hatte aber auch noch keine Störung. ;-)
Ich werde mal versuchen eine Störung zu provozieren. Dann kann ich sicher sagen, ob sich der Wert überhaupt ändert.

Viele Grüße
eMpTy

**Comment by empty88** [comment]
Ich konnte jetzt relativ einfach und schnell eine Störung erzeugen, indem ich den Stecker vom Speichertemperatursensor gezogen habe.
Und ja: obwohl die Fehlermeldung im Fehlerspeicher landet und auch am Gerät "Störung" angezeigt wurde, wurde keine 1 gesendet.
Habe auch mal eine andere Adresse probiert: 0x0847, keine Reaktion

In der [Viessmann Community](https://www.viessmann-community.com/t5/Sonstige-Produkte/VDensHO1-20C2-Auslesung-Stoerung-ueber-Optolink/td-p/172720) konnte man diesbezüglich wohl auch nicht weiterhelfen.

Da werde ich auch auf deine Methode wechseln. Leider ändert sich der Fehlerspeicher nicht, wenn die Störung verschwindet. Also lässt sich leider nicht aufzeichnen wie lange eine Störung bestand.

**Comment by empty88** [comment]
In Commit [a1df888](https://github.com/empty88/OptoLink/commit/a1df888d5dd72ac74ee9f107bc719124bdd77b22) habe ich jetzt auch deine Methode zur Fehlerbenachrichtigung umgesetzt.

**Comment by gitzila** [comment]
Moin, da hast Du wirklich Glück, dass Deine Heizung nicht so schlimm taktet wie meine. So sah es vor meiner "Manipulation" aus:

![2022-11-29-vorher](https://user-images.githubusercontent.com/49775197/210128227-67b7bf29-7939-41b7-97b9-84110691197f.jpg)

Und so jetzt:

![2022-11-29-nachher](https://user-images.githubusercontent.com/49775197/210128239-e9e4e045-ec45-4cc0-ae4c-cb9219c70020.jpg)

Frank

**Comment by gitzila** [comment]
Es ist schade, dass man den aktuellen Störungszustand nicht auslesen kann. Keine Ahnung, was die sich bei Viessmann da gedacht haben (wenn überhaupt). Immerhin bekommt man durch den Abgleich mit dem Protokoll überhaupt mit, dass etwas nicht stimmt. Ich habe mir für meinen ESP eine kleine Weboberfläche gebaut, die den letzten Fehler anzeigt und ich kann ihn dort manuell quittieren.

Andere Möglichkeiten, den aktuellen Fehlerzustand zu erkennen:

1. Mit einer zweiten Diode erkennen, ob die rote Störungs-LED leuchtet. Da müsste man aber etwas basteln, da der Platz ja physisch durch den Optolink belegt ist. Vielleicht eine Art Prisma, Spiegel o.ä.

2. Im Display das Störungssymbol / Text erkennen. Aber das ist im Vergleich zum Nutzen zu viel Aufwand.

3. Rein softwaretechnisch eine Heuristik umsetzen: Störung wurde erkannt, aber der Brenner läuft, Vorlauf wird warm, Pumpe läuft, WW wird warm -> also ist es wohl nicht so schlimm 😀.

Gefällt mir alles nicht so recht...

LG
Frank

**Comment by gitzila** [comment]
Hm, zu Lösung 1 fällt mir ein, dass in optischen Mäusen so ein Plastikding verbaut ist (glaub ich zumindest). So eine Art optischer "Splitter". Muss mal in meiner Wühlkiste suchen. Oder man lässt eine simple Fotozelle auf die LED "schielen". Es sollte sich ein Dauerrot im sichtbaren Bereich trotz Kommunikation im IR-Bereich erkennen lassen.

**Comment by gitzila** [comment]
Zur Vollständigkeit: Die HTML-Oberfläche für den ESP:

![Screenshot_2022-12-31-08-20-02-25_3aea4af51f236e4932235fdada7d1643](https://user-images.githubusercontent.com/49775197/210128826-fe556c8e-cc4c-46b8-b5c5-2171e9aa9944.jpg)

![Screenshot_2022-12-31-08-24-55-14_3aea4af51f236e4932235fdada7d1643](https://user-images.githubusercontent.com/49775197/210128957-5aefdc6d-553c-46f6-a6c4-26e382193bcb.jpg)

Der Button für "Fehler Quittieren" ist nicht vorhanden, da kein aktueller Fehler existiert. Ich lese nur den Protokolleintrag 1 aus, da das sonst alles zu lange dauert und meine Brennersteuerung aus dem Tritt kommt, die alle 2-3 Sekunden die Brennerleistung misst. Der Buffer der Optolink-Schnittstelle läuft dann leider über. 

Ich habe mir auch eine Berechnung integriert, die aus der Brennerleistung so ungefähr die verbrauchten m³ Gas herleitet. Ist so +-2% genau. Das reicht für meine Ansprüche hin.

**Reply by empty88** [reply]
Die Berechnung des Gasverbrauchs finde ich interessant, wo hast du die Werte her und wie hast du das Ergebnis überprüft?

**Reply by gitzila** [reply]
Relativ simpel: ich hole alle x Millisekunden (aktuell 2000) die Brennerleistung und vergleiche sie mit dem zuletzt gelesenen Wert. Das gibt 4 Möglichkeiten:

1. Vorher 0, nun > 0: Brenner gestartet
2. Vorher > 0, nun 0: Brenner gestoppt
3. Vorher 0, jetzt 0: Brenner ist und bleibt aus
4. Vorher > 0, jetzt > 0: Brenner in Betrieb

In den Fällen 1., 2. und 4. addiere ich die Brennerleistung und teile durch 2. Das ergibt einen ungefähren Durchschnittswert über die gemessene Zeit. Allerdings nehme ich nicht die fixen 2000 ms Intervall der Abfrage über den Optolink, sondern die tatsächlich vergangene Zeit zwischen den Messungen. Der ESP verdaddelt ja noch ein paar Millisekunden für andere Aufgaben, so dass i.d.R. so Werte um 2300 ms herauskommen.

Dann teile ich den Messwert der Leistung durch 2 (aus der Schnittstelle kommt ja der doppelte Wert laut Dokumentation). Das sind bei meiner Heizung dann Werte zwischen 16,5 und 65 Prozent. Der Brenner hat eine Nennleistung von 19.000 Watt. Zeit x Prozent x Nennleistung ergibt die verbrauchte Energie in Ws oder Joule. Das addiere ich laufend auf. 

Aus der Abrechnung des Versorgers habe ich den Brennwert und die Z-Zahl. Also die Joule teilen durch Brennwert und Z-Zahl. Und schon hat man die Kubikmeter.

Kommt allerdings überhaupt nicht hin. Der von mir berechnete Kubikmeter-Wert ist etwa 2,5* zu niedrig. Daher der Korrekturfaktor. Das lasse ich dann ein paar Tage laufen und vergleiche das mit der Gasuhr. 

Für einen ungefähren Verbrauchswert ist das ok für mich und besser als nix.

Warum allerdings die Werte um den offenbar festen Faktor 2,5 linear abweichen...? Keine Ahnung. Entweder ist das Gas gepanscht oder die Brennerleistung stimmt nicht oder die Nennleistung oder der Wirkungsgrad der Anlage ist Kacke.

Kann aber auch nicht sein, weil wir dieses Jahr mit 750m³ 144 qm Haus versorgt haben. Ein sehr guter Wert für ein EFH von 1999.

**Reply by empty88** [reply]
Danke für die detaillierte Beschreibung, aber das ist mir zu viel. Da werd ich wohl eher ein Projekt zum Auslesen der Gasuhr mal irgendwann angehen.
